File: packages/lynx-file-system/android/com/sigx/filesystem/FileSystemModule.kt
package com.sigx.filesystem

import android.content.Context
import android.net.Uri
import android.util.Base64
import android.util.Log
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.react.bridge.Callback
import com.lynx.react.bridge.JavaOnlyMap
import java.io.File

/**
 * File system access module.
 * JS usage: NativeModules.FileSystem.readFile("data.json", callback)
 */
class FileSystemModule(context: Context) : LynxModule(context) {

    companion object {
        private const val TAG = "FileSystemModule"
    }

    @LynxMethod
    fun readFile(path: String?, callback: Callback?) {
        try {
            val file = resolveFile(path ?: "")
            if (!file.exists()) {
                val error = JavaOnlyMap()
                error.putString("error", "File not found: $path")
                callback?.invoke(error)
                return
            }
            callback?.invoke(file.readText())
        } catch (e: Exception) {
            val error = JavaOnlyMap()
            error.putString("error", e.message ?: "Unknown error")
            callback?.invoke(error)
        }
    }

    /**
     * Read a file as raw bytes, returned base64-encoded. Accepts the same
     * paths as `readFile`, plus `file://` URIs and `content://` URIs —
     * i.e. anything a picker hands back (`content://` grants are read via
     * the ContentResolver, not the File API).
     */
    @LynxMethod
    fun readFileBase64(path: String?, callback: Callback?) {
        if (path.isNullOrEmpty()) {
            val error = JavaOnlyMap()
            error.putString("error", "Path is required")
            callback?.invoke(error)
            return
        }
        try {
            val p = path
            val bytes: ByteArray? = if (p.startsWith("content://")) {
                mContext.contentResolver.openInputStream(Uri.parse(p))?.use { it.readBytes() }
            } else {
                val file = resolveFile(p)
                if (!file.exists()) {
                    val error = JavaOnlyMap()
                    error.putString("error", "File not found: $path")
                    callback?.invoke(error)
                    return
                }
                file.readBytes()
            }
            if (bytes == null) {
                val error = JavaOnlyMap()
                error.putString("error", "Could not open stream: $path")
                callback?.invoke(error)
                return
            }
            callback?.invoke(Base64.encodeToString(bytes, Base64.NO_WRAP))
        } catch (e: Exception) {
            val error = JavaOnlyMap()
            error.putString("error", e.message ?: "Unknown error")
            callback?.invoke(error)
        }
    }

    @LynxMethod
    fun writeFile(path: String?, content: String?, callback: Callback?) {
        try {
            val file = resolveFile(path ?: "")
            file.parentFile?.mkdirs()
            file.writeText(content ?: "")
            callback?.invoke(true)
        } catch (e: Exception) {
            val error = JavaOnlyMap()
            error.putString("error", e.message ?: "Unknown error")
            callback?.invoke(error)
        }
    }

    @LynxMethod
    fun deleteFile(path: String?, callback: Callback?) {
        try {
            val p = path ?: ""
            // `content://` is a ContentResolver grant, not a path: `File()`
            // would build a nonexistent relative file under filesDir, whose
            // delete "succeeds" while the picked document is untouched. The
            // app doesn't own that file — say so instead of reporting a
            // deletion that never happened.
            if (p.startsWith("content://")) {
                val error = JavaOnlyMap()
                error.putString(
                    "error",
                    "Cannot delete a content:// URI — the app doesn't own that file: $p",
                )
                callback?.invoke(error)
                return
            }
            val file = resolveFile(p)
            // `delete()` returns false both for "already gone" — the documented
            // no-op, matching iOS — and for a real failure: a non-empty
            // directory, or a read-only path. Only the latter is an error.
            if (!file.delete() && file.exists()) {
                val error = JavaOnlyMap()
                error.putString("error", "Failed to delete: $p")
                callback?.invoke(error)
                return
            }
            callback?.invoke(true)
        } catch (e: Exception) {
            val error = JavaOnlyMap()
            error.putString("error", e.message ?: "Unknown error")
            callback?.invoke(error)
        }
    }

    @LynxMethod
    fun getInfo(path: String?, callback: Callback?) {
        try {
            val file = resolveFile(path ?: "")
            val result = JavaOnlyMap()
            result.putString("uri", file.absolutePath)
            result.putDouble("size", file.length().toDouble())
            result.putBoolean("exists", file.exists())
            result.putBoolean("isDirectory", file.isDirectory)
            result.putDouble("modifiedAt", file.lastModified().toDouble())
            callback?.invoke(result)
        } catch (e: Exception) {
            val error = JavaOnlyMap()
            error.putString("error", e.message ?: "Unknown error")
            callback?.invoke(error)
        }
    }

    @LynxMethod
    fun getDocumentDirectory(): String {
        return mContext.filesDir.absolutePath
    }

    @LynxMethod
    fun getCacheDirectory(): String {
        return mContext.cacheDir.absolutePath
    }

    private fun resolveFile(path: String): File {
        // Accept `file://` URIs (pickers return these) alongside plain paths.
        val p = if (path.startsWith("file://")) Uri.parse(path).path ?: path else path
        val f = File(p)
        return if (f.isAbsolute) f else File(mContext.filesDir, p)
    }
}

File: packages/lynx-file-system/ios/FileSystemModule.swift
import Foundation
import Lynx

/// File system access module.
/// JS usage: NativeModules.FileSystem.readFile("data.json", callback)
class FileSystemModule: NSObject, LynxModule {

    @objc static var name: String { "FileSystem" }

    @objc static var methodLookup: [String: String] {
        [
            "readFile": NSStringFromSelector(#selector(readFile(_:callback:))),
            "readFileBase64": NSStringFromSelector(#selector(readFileBase64(_:callback:))),
            "writeFile": NSStringFromSelector(#selector(writeFile(_:content:callback:))),
            "deleteFile": NSStringFromSelector(#selector(deleteFile(_:callback:))),
            "getInfo": NSStringFromSelector(#selector(getInfo(_:callback:))),
            "getDocumentDirectory": NSStringFromSelector(#selector(getDocumentDirectory)),
            "getCacheDirectory": NSStringFromSelector(#selector(getCacheDirectory)),
        ]
    }

    required override init() { super.init() }
    required init(param: Any) { super.init() }

    private let fileManager = FileManager.default

    @objc func readFile(_ path: String?, callback: LynxCallbackBlock?) {
        guard let path = path else {
            callback?(["error": "Path is required"])
            return
        }
        let resolvedPath = resolveFile(path)
        guard fileManager.fileExists(atPath: resolvedPath) else {
            callback?(["error": "File not found: \(path)"])
            return
        }
        do {
            let content = try String(contentsOfFile: resolvedPath, encoding: .utf8)
            callback?(content)
        } catch {
            callback?(["error": error.localizedDescription])
        }
    }

    /// Read a file as raw bytes, returned base64-encoded. Accepts the same
    /// paths as `readFile` plus `file://` URIs (what pickers hand back).
    @objc func readFileBase64(_ path: String?, callback: LynxCallbackBlock?) {
        guard let path = path else {
            callback?(["error": "Path is required"])
            return
        }
        let resolvedPath = resolveFile(path)
        guard fileManager.fileExists(atPath: resolvedPath) else {
            callback?(["error": "File not found: \(path)"])
            return
        }
        do {
            let data = try Data(contentsOf: URL(fileURLWithPath: resolvedPath))
            callback?(data.base64EncodedString())
        } catch {
            callback?(["error": error.localizedDescription])
        }
    }

    @objc func writeFile(_ path: String?, content: String?, callback: LynxCallbackBlock?) {
        guard let path = path else {
            callback?(["error": "Path is required"])
            return
        }
        let resolvedPath = resolveFile(path)
        let directory = (resolvedPath as NSString).deletingLastPathComponent
        try? fileManager.createDirectory(atPath: directory, withIntermediateDirectories: true)
        do {
            try (content ?? "").write(toFile: resolvedPath, atomically: true, encoding: .utf8)
            callback?(true)
        } catch {
            callback?(["error": error.localizedDescription])
        }
    }

    @objc func deleteFile(_ path: String?, callback: LynxCallbackBlock?) {
        guard let path = path else {
            callback?(["error": "Path is required"])
            return
        }
        let resolvedPath = resolveFile(path)
        // Deleting something that isn't there is the documented no-op, and what
        // Android's `File.delete()` does — the JS side now throws on `{ error }`,
        // so `removeItem`'s "no such file" must not reach it.
        //
        // Decided by ATTEMPTING the delete and reading the failure, not by an
        // existence pre-check: `fileExists(atPath:)` **traverses symlinks**, so
        // a dangling link reads as absent and would be skipped — left on disk,
        // where the unguarded `removeItem` used to remove the link itself. The
        // same blindness hides a permission-denied path as a silent success.
        do {
            try fileManager.removeItem(atPath: resolvedPath)
            callback?(true)
        } catch let error as NSError {
            // ENOENT, however it is reported: nothing to delete is not a
            // failure, and this also covers losing the race with another
            // deleter. Anything else is real and must reach the caller.
            let isMissing = (error.domain == NSCocoaErrorDomain && error.code == NSFileNoSuchFileError)
                || (error.domain == NSPOSIXErrorDomain && error.code == Int(ENOENT))
            if isMissing {
                callback?(true)
            } else {
                callback?(["error": error.localizedDescription])
            }
        }
    }

    @objc func getInfo(_ path: String?, callback: LynxCallbackBlock?) {
        guard let path = path else {
            callback?(["error": "Path is required"])
            return
        }
        let resolvedPath = resolveFile(path)
        let exists = fileManager.fileExists(atPath: resolvedPath)
        var result: [String: Any] = ["uri": resolvedPath, "exists": exists]
        if exists, let attrs = try? fileManager.attributesOfItem(atPath: resolvedPath) {
            result["size"] = (attrs[.size] as? UInt64) ?? 0
            result["isDirectory"] = (attrs[.type] as? FileAttributeType) == .typeDirectory
            result["modifiedAt"] = (attrs[.modificationDate] as? Date)?.timeIntervalSince1970 ?? 0
        }
        callback?(result)
    }

    @objc func getDocumentDirectory() -> String {
        NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first ?? ""
    }

    @objc func getCacheDirectory() -> String {
        NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true).first ?? ""
    }

    private func resolveFile(_ path: String) -> String {
        // Accept `file://` URIs (pickers return these) alongside plain paths.
        if path.hasPrefix("file://") {
            if let url = URL(string: path) { return url.path }
            // URL(string:) rejects unescaped characters (e.g. spaces in a
            // JS-prefixed bare path) — strip the scheme manually instead of
            // falling through to relative-path resolution.
            let stripped = String(path.dropFirst("file://".count))
            return stripped.removingPercentEncoding ?? stripped
        }
        if (path as NSString).isAbsolutePath { return path }
        let docsDir = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first ?? ""
        return (docsDir as NSString).appendingPathComponent(path)
    }
}
